### DISCUSSION TITLE
Create Astro Template Outlet

### DISCUSSION THREAD

**Discussion by louiss0** [withastro/roadmap#364] [Proposal]
- Start Date: 2022/10/25
- Reference Issues: 
- Implementation PR: 

# Summary

I created a component  called `<AstroTemplateOutlet />` it's a component that takes the content of a template and replaces it's content with that of the content from a template. I think it should be added to set of components called 
`astro/components`. If added developers could render HTML from templates easily without having to resort to finding a package or giving up and having to create a component. Giving developers the power to render the content of HTML templates saves developers time and is sometimes more ergonomic than creating new file's. I call this component `<AstroTemplateOutlet />` but you guys can feel free to call it whatever you want.  

# Motivation

People including myself like to store small chunks of HTML inside of a file and display them in the final output. Of a template in Vue or React. Frameworks that support SFC's don't ever support the creation of  small chunks of HTML inside of templates at all. In the Vue world people download a package called "vue-plugin-jsx". That package solves their problem but they can't use directives at all. Astro doesn't support this mechanism it doesn't allow you to specify a different template format at all. Someone decided to bringing this up so an concept called named templates was created  [Named Templates ](https://vue-macros.sxzz.moe/features/named-template.html). Angular has this concept as well [NG Template Outlet ](https://angular.io/api/common/NgTemplateOutlet)  

As you can see the need to create bits of HTML in a single file then bringing them back together in the same file is in huge demand, and **this is a problem that people will ask Astro to solve at some point.** The problem with creating components is that you have to create a new file for each component every time.  **You want to split small chunks of HTML into pieces and reassemble them back in a reasonable way so that individual pieces of HTML can be edited and focused on without creating another file.**     

>[!info]  **My main point is that creating a files all the time is simply frustrating and pointless.**
>- Small files are harder to create, find, and change.
>- Creating components for a bit of HTML that you know will only be used once is a waste of time. 
>- Well named chunks of HTML in a file is easier to find delete and edit. 

# Detailed design

When you import and then use `<AstroTemplateOutlet/>` a script will be injected into the page that searches for templates with  an `id=` and a the element rendered by `<AstroTemplateOutlet />` with a `data-template-name=` attribute. After that it tries to take the content of a template with a specified `id=`  and replaces the HTML rendered by the outlet with the one that is in the template. `<AstroTemplateOutlet />`  requires and `templateId=` prop. That prop is the id of a template.  The prop is then used to give the HTML rendered by the outlet an attribute of  `data-template-name=`  with the prop passed as the value.           

### Test app Example 

**Writing this HTML**
```html

<template id="template-one">
	This is my template
</template>

<div>
<AstroTemplateOutlet templateId="template-one" /> 
</div>

```

**Will render this html**

```html

<template id="template-one">
	This is my template
</template>

<div>
This is my template 
</div>

```

# Testing strategy

When it comes to testing this component it's better to use integration testing. You want to see how it works on a page. Cypress is mature and can be used for just that purpose. Using the official astro component test is not good. This is a component that relies on an existing page. I created a repo with the source code there You can take the source code fork it and test it there [Template Outlet Source Code](https://github.com/louiss0/astro-template-outlet)  

I think you should use [Cypress End To End Testing](https://docs.cypress.io/guides/end-to-end-testing/writing-your-first-end-to-end-test)
# Drawbacks

Other than developers having to learn a new syntax. Astro only allows you to store templates in tags or at the bottom of a file. Developers will have to put templates at the bottom of the file if they want to use them. `<AstroTemplateOutlet />` doesn't allow you to pass a slot to it at all or anything else at the moment so it will not be able to do much with templates. 

# Alternatives

The impact of not doing this is that users of astro will not be able to separate their JSX expressions or small bits of HTML into pieces that they can use inside of the template. Instead of doing this you could reconsider not allowing JSX to be used inside of the server frontmatter. That mechanism is a stronger one than what I'm proposing.  

# Adoption strategy

Please consider:

- If we implement this proposal, how will existing Astro developers adopt it?
	- If this proposal is added to core it should be able to be accessed from the "astro/components" package and will be activated there. 
	- Developers could then remove most of the folders they used to co-locate components and instead store them inside of templates instead. 
	- VS Code allows you to search for HTML elements by using an id templates can be easily found that way.
- Is this a breaking change? Can we write a codemod?
	- No
	- I don't know about codemods but I did publish this on npm  
- Can we provide a runtime adapter library for the original API it replaces?
	- It affect's no other libraries 
- How will this affect other projects in the Astro ecosystem? 
	- It will not affect them at all they should still be fine.  

# Unresolved questions

Optional, but suggested for first drafts.
What parts of the design are still to be determined?

There could be a way of passing props into the template that is being rendered. 

```jsx

<template name="user-template">
{(context)=>  context.name }
</template>

<AstroTemplateOutlet 
templateName="user-template" 
context={{name:"James",}}
/>   

```

I don't know enough Astro internals to determine if this is possible. I'd like it if you did do this for me. Thank You.

**Comment by lloydjatkinson** [comment]
>  It also doesn't support putting bits of HTML into separate pieces to be used.

> I believe that providing an API for users to take bits of HTML and then put them back together in a declarative way

Both of these are achieved with components

**Reply by louiss0** [reply]
Astro does not support the creation and invoking of components from the same file. You don't want to always put pieces of HTML in different files. Sometimes you want to split them apart and give them a name then put them back together in an organized way. Having to create a new file every time you want to make a component is pure overhead

**Reply by louiss0** [reply]
# Addressing  Concerns 

I have heard from the core team that they see no point in implementing this RFC because this is just another way of creating a component. While this is true. If you decide not to send in a context **Yes!** it would just be an inferior way of creating components. **The main point of this is that developers don't always have to create files in order to take sections of HTML name them and break them up into pieces and create a file for all of them**.  **They can instead just take the pieces of HTML split them apart and then reassemble them back to where they need to be in without having to create a file**. 

## Creating A Component 

When it comes to separating huge chunks of HTML into pieces the developer doesn't have to create a file for each chunk of HTML that needs to be created. Creating a file for separate chunks of HTML is a waste of time. It makes the developer burn time and it makes them waste too much time trying to find files that have nothing to do with the project as a whole but only for one page or layout or even a component.  

>[!info]  Allowing the developer to just use templates to structure and reassemble their HTML. does the following things.
> - It allows the developer to get rid of HTML quickly when requirements change 
> - It allows the developer to focus on many pieces of HTML without having to go from one file to another allowing them to save time
 
 
## You don't need to implement context 

I suggested implementing a context so that a variable could be passed into the context. You don't have to do that at all. I just thought it would be a good thing. 
It would allow developers to just make reusable components without having to create a file. But I don't need that at all and most developers would just create a component at that point. 


## You said you wanted examples of this well you got them 

Many members asked for this in the Vue community because when you use JSX you have to write imperative code and have no access to the power of directives and can't get good typing for slots at all. The idea for a context came from the way Angular does it. 

- Angular Template Implementation 
	- https://angular.io/api/common/NgTemplateOutlet
- Vue Template Experimental  Implementation 
	- https://vue-macros.sxzz.moe/features/named-template


## How I use Astro Template Outlet In My Recent Project

The examples below are from a website that I'm cloning. These are templates that are created and then put into the HTML The first five pictures are templates that I created. The last picture is where I assembled all of the templates. I think this example might spark some inspiration.   


![copyright-block-template](https://user-images.githubusercontent.com/29028601/204667835-a0b05fe7-5b60-4a70-9f98-f94dfbd7823d.png)

![tag-links-template](https://user-images.githubusercontent.com/29028601/204667867-2bbc65e2-7d5d-4c70-9c64-6d290cf164d9.png)

![category-links-template](https://user-images.githubusercontent.com/29028601/204667904-79af02d5-4d5b-42fc-960f-1aa71fb7c0eb.png)

![social-links-template](https://user-images.githubusercontent.com/29028601/204667932-31a07588-8685-4c62-85b8-99e359c57165.png)

![menu-links-template](https://user-images.githubusercontent.com/29028601/204668015-b5718ee8-148a-4cf1-be82-4af46b47db56.png)


![menus-area](https://user-images.githubusercontent.com/29028601/204668031-dcd74062-7236-4aa1-8f33-18ebca0e88af.png)


## Naming 

I don't care about what the name is as long it communicates what the component is doing. Outlet is used to say where something is going to be rendered. I think that's why I used it. I'm just asking you to add the component into core so that I and other developers can use it without having to download it from npm.

**Comment by louiss0** [comment]
# Addressing  Concerns 

I have heard from the core team that they see no point in implementing this RFC because this is just another way of creating a component. While this is true. If you decide not to send in a context **Yes!** it would just be an inferior way of creating components. **The main point of this is that developers don't always have to create files in order to take sections of HTML name them and break them up into pieces and create a file for all of them**.  **They can instead just take the pieces of HTML split them apart and then reassemble them back to where they need to be in without having to create a file**. 

## Creating A Component 

When it comes to separating huge chunks of HTML into pieces the developer doesn't have to create a file for each chunk of HTML that needs to be created. Creating a file for separate chunks of HTML is a waste of time. It makes the developer burn time and it makes them waste too much time trying to find files that have nothing to do with the project as a whole but only for one page or layout or even a component.  

>[!info]  Allowing the developer to just use templates to structure and reassemble their HTML. does the following things.
> - It allows the developer to get rid of HTML quickly when requirements change 
> - It allows the developer to focus on many pieces of HTML without having to go from one file to another allowing them to save time
 
 
## You don't need to implement context 

I suggested implementing a context so that a variable could be passed into the context. You don't have to do that at all. I just thought it would be a good thing. 
It would allow developers to just make reusable components without having to create a file. But I don't need that at all and most developers would just create a component at that point. 


## You said you wanted examples of this well you got them 

Many members asked for this in the Vue community because when you use JSX you have to write imperative code and have no access to the power of directives and can't get good typing for slots at all. The idea for a context came from the way Angular does it. 

- Angular Template Implementation 
	- https://angular.io/api/common/NgTemplateOutlet
- Vue Template Experimental  Implementation 
	- https://vue-macros.sxzz.moe/features/named-template


## How I use Astro Template Outlet In My Recent Project

The examples below are from a website that I'm cloning. These are templates that are created and then put into the HTML The first five pictures are templates that I created. The last picture is where I assembled all of the templates. I think this example might spark some inspiration.   


![copyright-block-template](https://user-images.githubusercontent.com/29028601/204667835-a0b05fe7-5b60-4a70-9f98-f94dfbd7823d.png)

![tag-links-template](https://user-images.githubusercontent.com/29028601/204667867-2bbc65e2-7d5d-4c70-9c64-6d290cf164d9.png)

![category-links-template](https://user-images.githubusercontent.com/29028601/204667904-79af02d5-4d5b-42fc-960f-1aa71fb7c0eb.png)

![social-links-template](https://user-images.githubusercontent.com/29028601/204667932-31a07588-8685-4c62-85b8-99e359c57165.png)

![menu-links-template](https://user-images.githubusercontent.com/29028601/204668015-b5718ee8-148a-4cf1-be82-4af46b47db56.png)


![menus-area](https://user-images.githubusercontent.com/29028601/204668031-dcd74062-7236-4aa1-8f33-18ebca0e88af.png)


## Naming 

I don't care about what the name is as long it communicates what the component is doing. Outlet is used to say where something is going to be rendered. I think that's why I used it. I'm just asking you to add the component into core so that I and other developers can use it without having to download it from npm.

**Reply by delucis** [reply]
Thanks for sharing @louiss0! It sounds like you’re able to use this component in your projects already without this being built into Astro. In that case, might it make sense to publish as a user project on npm rather than making it a part of core Astro? We have [a guide to publishing](https://docs.astro.build/en/reference/publish-to-npm/) in the Astro docs, including [instructions on how to configure your package](https://docs.astro.build/en/reference/publish-to-npm/#integrations-library) so it appears in the [Astro integrations catalogue](https://astro.build/integrations/).

**Reply by louiss0** [reply]
> Thanks for sharing @louiss0! It sounds like you’re able to use this component in your projects already without this being built into Astro. In that case, might it make sense to publish as a user project on npm rather than making it a part of core Astro? We have [a guide to publishing](https://docs.astro.build/en/reference/publish-to-npm/) in the Astro docs, including [instructions on how to configure your package](https://docs.astro.build/en/reference/publish-to-npm/#integrations-library) so it appears in the [Astro integrations catalogue](https://astro.build/integrations/).

I decided to add it as an integration. Is there any way for it to be added to the official integrations page for Astro. I sent it along with another component that I created called `astro-gap`. I even showcased both on discord. I [Astro Gap](https://www.npmjs.com/package/astro-gap) [Astro Template Outlet](https://www.npmjs.com/package/astro-template-outlet). 

I hope these get added to the integrations showcase.

**Reply by delucis** [reply]
Yes, you can follow the instructions here: https://docs.astro.build/en/reference/publish-to-npm/#integrations-library

TL;DR — include `"astro-component"` in the `keywords` array in `package.json` and the integration library will pick up your package automatically when it updates (I think updates are once per day?)

**Reply by delucis** [reply]
Oh, I see you added that keyword already, but I think you also need a category keyword, e.g. `"ui"` to be included in one of the categories on [astro.build](https://astro.build/integrations/).

**Reply by louiss0** [reply]
> I added UI as a keyword thanks for the help. I think I'll work on another component called NavLink now I need a `<Link />` that can be used without having to use a specific template in Astro.
